Makefile

```
@@ -1,51 +1,92 @@
-BUF_BIN="`pwd`/bin"
-BUF_TARGET?=buf.build/viamrobotics/api --path common,component,robot,service
-NPM_BIN="`pwd`/node_modules/.bin"
-PATH_WITH_TOOLS="${BUF_BIN}:${NPM_BIN}:${PATH}"
+buf := ./bin/buf
+node_modules := $(shell npm root)
 
-clean-dist:
-	rm -rf dist
+export PATH := $(node_modules)/.bin:$(PATH)
 
-install:
-	npm ci --audit=false
+# common targets
+
+.PHONY: all
+all: clean build lint test
+
+.PHONY: setup
+setup: setup-buf setup-js
+
+.PHONY: teardown
+teardown: teardown-buf teardown-js
 
-test:
+.PHONY: build
+build: build-buf build-js
+
+.PHONY: clean
+clean: clean-buf clean-js
+
+.PHONY: test
+test: $(node_modules)
 	npm run test
 
-lint: install
+.PHONY: lint
+lint: $(node_modules)
 	npm run lint:prettier
 	npm run lint:eslint
 	npm run typecheck
 	npm run check
 
-format: install
+.PHONY: format
+format: $(node_modules)
 	npm run format
 
-buf-clean:
-	rm -rf src/gen/
+# protobuf targets
 
+.PHONY: setup-buf
+setup-buf: $(buf)
 
-buf-install: buf-clean
-	./etc/install_buf.sh $(BUF_BIN)
+$(buf):
+	./etc/install_buf.sh ./bin
 
-buf-update: buf-install
-	PATH=$(PATH_WITH_TOOLS) buf mod update
+.PHONY: teardown-buf
+teardown-buf:
+	rm -rf $(buf)
 
-buf: install buf-install
-	PATH=$(PATH_WITH_TOOLS) buf generate buf.build/googleapis/googleapis
-	PATH=$(PATH_WITH_TOOLS) buf generate ${BUF_TARGET}
-	PATH=$(PATH_WITH_TOOLS) buf generate buf.build/erdaniels/gostream
-	PATH=$(PATH_WITH_TOOLS) buf generate buf.build/viamrobotics/goutils
+.PHONY: clean-buf
+clean-buf:
+	rm -rf src/gen
 
-build: clean-dist buf
-	# TODO(RSDK-870): try removing the custom `--max-old-space-size` option
-	# once we migrate to protobuf-es, since that generator should produce
-	# much smaller javascript bundles.
-	NODE_OPTIONS="--max-old-space-size=16384" npm run build
+.PHONY: update-buf
+update-buf: $(buf)
+	$(buf) mod update
+
+.PHONY: build-buf
+build-buf: $(buf) $(node_modules) clean-buf
+	$(buf) generate buf.build/googleapis/googleapis
+	$(buf) generate buf.build/viamrobotics/api --path common,component,robot,service
+	$(buf) generate buf.build/erdaniels/gostream
+	$(buf) generate buf.build/viamrobotics/goutils
 
-# LOCAL DEVELOPMENT HELPERS
+# js targets
+
+.PHONY: setup-js
+setup-js: $(node_modules)
+
+$(node_modules): package-lock.json
+	npm ci --audit=false
+
+.PHONY: teardown-js
+teardown-js:
+	rm -rf node_modules
+
+.PHONY: clean-js
+clean-js:
+	rm -rf dist
+
+# TODO(RSDK-870): try removing the custom `--max-old-space-size` option
+# once we migrate to protobuf-es, since that generator should produce
+# much smaller javascript bundles.
+.PHONY: build-js
+build-js: $(node_modules) clean-js build-buf
+	NODE_OPTIONS="--max-old-space-size=16384" npm run build
```

```suggestion
.PHONY: build-js
build-js: $(node_modules) clean-js build-buf
	npm run build
```

another thing we can get rid of, thanks to https://github.com/viamrobotics/viam-typescript-sdk/pull/41